### PR TITLE
[PLAT-77308] New aggregation rule filter separator

### PR DESCRIPTION
**Pull request by hczhu**
This PR introduces another aggregation rule filter separator, `#`. It has a high priority than the default one, `:`.
Added unit test cases.

```
$ go test -timeout 30s github.com/m3db/m3/src/metrics/filters

```

```
$ make m3coordinator
```
Test rules
```
    - "drop": true
      "filter": "__name__#common:container:cpu_throttling_percent:10m kubernetes_namespace:m3"
      "name": "blockingMappingRule: common:container:cpu_throttling_percent:10m"
    - "drop": true
      "filter": "__name__#common:container kubernetes_namespace#m3"
      "name": "blockingMappingRule: common:container"
```
```
$ bin/m3coordinator -f ~/data/m3coord.yaml -z

2023/03/21 09:56:53 the config is valid.
```

--
```
    - "drop": true
      "filter": "__name__#common:container:cpu_throttling_percent:10m#10m kubernetes_namespace:m3"
      "name": "blockingMappingRule: common:container:cpu_throttling_percent:10m"
```
```
$ bin/m3coordinator -f ~/data/m3coord.yaml -z

2023/03/21 09:57:38 error validating config: Error from adding mapping rule [blockingMappingRule: common:container:cpu_throttling_percent:10m]: cannot add rule blockingMappingRule: common:container:cpu_throttling_percent:10m: tags filter __name__#common:container:cpu_throttling_percent:10m#a kubernetes_namespace:m3 is malformed: invalid filter __name__#common:container:cpu_throttling_percent:10m#a: expecting tag pattern pairs

```

### CODE REVIEW

File: src/metrics/filters/tags_filter.go [hunk 1]
```diff
@@ -39,11 +39,18 @@ var (
...
+	}
+	databricksFilterSeparator = tagFilterSeparator{
+		str: "#", negate: false,
+	}
 
 	// validFilterSeparators represent a list of valid filter separators.
 	// NB: the separators here are tried in order during parsing.
 	validFilterSeparators = []tagFilterSeparator{
+		// The separators are sorted by priorities here. If one applies successfully, the ones below it are not considered.
+		databricksFilterSeparator,
```
Comment (davidyuanfs) [inline src/metrics/filters/tags_filter.go:53]: It is smart to add a new separator to replace `:`, one concern, promSQL use '#' as [comment](https://prometheus.io/docs/prometheus/latest/querying/basics/#comments), but I am thinking it shouldn't have conflict with our rule definition?

Comment (jnyi) [reply src/metrics/filters/tags_filter.go:53]: will m3 filter based on `:` correctly? Can we do a unit test on that?

Comment (hczhu) [reply src/metrics/filters/tags_filter.go:53]: PromQL comment is irrelevant to aggregation rules.

> will m3 filter based on : correctly? Can we do a unit test on that?

Yes, it's covered by the existing unit tests.

Comment (hczhu) [reply src/metrics/filters/tags_filter.go:53]: > add a new separator to replace :

It's not to replace ':'. The logic here is to separate by '#' first. If it fails, try ':' then. Why do we try '#' first? It's to solve the following ambiguous case.
```
__name__#metric_name:10m
```
If we try ':' first, the case above will be parsed incorrectly.

Comment (hczhu) [reply src/metrics/filters/tags_filter.go:53]: '#' is not an acceptable char for metric names and label names/values. Therefore, this change won't impact any existing aggregation rules.

image

Comment (davidyuanfs) [reply src/metrics/filters/tags_filter.go:53]: sg, this is awesome!

**Review by davidyuanfs — APPROVED** [review]
lgtm

**Review by jnyi — APPROVED** [review]
LGTM, nice

**Pull request merged**